File: languages/go/liner/src/Lexer/Lexer.php
<?php

declare(strict_types=1);

namespace Serhii\Liner\Lexer;

use Serhii\Liner\Token\Token;
use Serhii\Liner\Token\TokenType;

class Lexer
{
    private int $position = 0;
    private int $nextPosition = 0;
    private string $input;
    private string $char;

    public function __construct(string $input)
    {
        $this->input = $input;
        $this->advanceChar();
    }

    public function nextToken(): Token
    {
        $this->skipWhitespace();

        $token = null;

        if ($this->char === '"') {
            $token = new Token(TokenType::STR, $this->readString());
        } else if ($this->char === '.') {
            $token = new Token(TokenType::PERIOD, $this->char);
        } else if ($this->isLetter()) {
            $ident = $this->readIdentifier();
            return new Token(TokenType::lookupIdentifier($ident), $ident);
        } else if ($this->isNumber()) {
            return new Token(TokenType::INT, $this->readNumber());
        } else if ($this->char === '(') {
            $token = new Token(TokenType::LBRACE, $this->char);
        } else if ($this->char === ')') {
            $token = new Token(TokenType::RBRACE, $this->char);
        } else if ($this->char === '') {
            $token = new Token(TokenType::EOF, '');
        } else {
            $token = new Token(TokenType::ILLEGAL, $this->char);
        }

        $this->advanceChar();

        return $token;
    }

    private function skipWhitespace(): void
    {
        while ($this->char === ' ' || $this->char === "\n" || $this->char === "\t" || $this->char === "\r") {
            $this->advanceChar();
        }
    }

    private function advanceChar(): void
    {
        if ($this->nextPosition >= strlen($this->input)) {
            $this->char = '';
        } else {
            $this->char = $this->input[$this->nextPosition];
        }

        $this->position = $this->nextPosition;
        $this->nextPosition += 1;
    }

    private function peekChar(): string
    {
        if ($this->nextPosition >= strlen($this->input)) {
            return '';
        }

        return $this->input[$this->nextPosition];
    }

    private function isLetter(): bool
    {
        return ctype_alpha($this->char);
    }

    private function isNumber(): bool
    {
        return is_numeric($this->char);
    }

    private function readIdentifier(): string
    {
        $start = $this->position;

        while ($this->isLetter()) {
            $this->advanceChar();
        }

        return substr($this->input, $start, $this->position - $start);
    }

    private function readNumber(): string
    {
        $start = $this->position;

        while ($this->isNumber()) {
            $this->advanceChar();
        }

        return substr($this->input, $start, $this->position - $start);
    }

    private function readString(): string
    {
        $start = $this->position + 1;

        while (true) {
            $prev = $this->char;

            $this->advanceChar();

            if (($this->char === '"' || $this->char === '') && $prev != '\\') {
                break;
            }
        }

        return substr($this->input, $start, $this->position - $start);
    }
}

File: languages/go/marf-lang/lexer/lexer.go
package lexer

import (
	"marf/token"
)

type Lexer struct {
	input        string
	position     int
	nextPosition int
	char         byte
}

func New(inp string) *Lexer {
	l := &Lexer{input: inp}
	l.advanceChar()

	return l
}

func (l *Lexer) NextToken() token.Token {
	var tok token.Token

	l.skipWhitespace()

	switch l.char {
	case ';':
		tok = newToken(token.SEMI, ";")
	case '+':
		tok = newToken(token.PLUS, "+")
	case '[':
		tok = newToken(token.LBRACKET, "[")
	case ']':
		tok = newToken(token.RBRACKET, "]")
	case '=':
		if l.peekChar() == '=' {
			l.advanceChar()
			tok = newToken(token.EQ, "==")
		} else {
			tok = newToken(token.ASSIGN, "=")
		}
	default:
		if isLetter(l.char) {
			tok.Literal = l.readIdentifier()
			tok.Type = token.LookupIdentifier(tok.Literal)
			return tok
		}

		if isDigit(l.char) {
			tok.Literal = l.readNumber()
			tok.Type = token.INT
			return tok
		}
	}

	l.advanceChar()

	return tok
}

func (l *Lexer) peekChar() byte {
	if l.nextPosition >= len(l.input) {
		return 0
	}

	return l.input[l.nextPosition]
}

func (l *Lexer) skipWhitespace() {
	for l.char == ' ' || l.char == '\n' || l.char == '\t' || l.char == '\r' {
		l.advanceChar()
	}
}

func (l *Lexer) readIdentifier() string {
	position := l.position

	for isLetter(l.char) {
		l.advanceChar()
	}

	return l.input[position:l.position]
}

func (l *Lexer) readNumber() string {
	position := l.position

	for isDigit(l.char) {
		l.advanceChar()
	}

	return l.input[position:l.position]
}

func (l *Lexer) advanceChar() {
	if l.nextPosition >= len(l.input) {
		l.char = 0
	} else {
		l.char = l.input[l.nextPosition]
	}

	l.position = l.nextPosition
	l.nextPosition += 1
}

func isLetter(char byte) bool {
	return (char >= 'a' && char <= 'z') || (char >= 'A' && char <= 'Z') || char == '_'
}

func isDigit(char byte) bool {
	return char >= '0' && char <= '9'
}

func newToken(tokenType token.TokenType, literal string) token.Token {
	return token.Token{
		Type:    tokenType,
		Literal: literal,
	}
}
